File: 2025-01-January-GFG-POTD/January 14 - Container With Most Water/Solution.java
//{ Driver Code Starts
import java.io.*;
import java.util.*;

class Sorting {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());
        for (int g = 0; g < t; g++) {
            String[] str = (br.readLine()).trim().split(" ");
            int arr[] = new int[str.length];
            for (int i = 0; i < str.length; i++) arr[i] = Integer.parseInt(str[i]);
            System.out.println(new Solution().maxWater(arr));
            System.out.println("~");
        }
    }
}
// } Driver Code Ends


// User function Template for Java

class Solution {

    public int maxWater(int arr[]) {
        // Code Here
        int left = 0; // Left pointer
        int right = arr.length - 1; // Right pointer
        int maxWater = 0; // To store the maximum water

        while (left < right) {
            // Calculate the water between the current pair of lines
            int height = Math.min(arr[left], arr[right]);
            int width = right - left;
            int currentWater = height * width;

            // Update the maximum water if the current one is greater
            maxWater = Math.max(maxWater, currentWater);

            // Move the pointer pointing to the shorter line inward
            if (arr[left] < arr[right]) {
                left++;
            } else {
                right--;
            }
        }

        return maxWater;
    }
}

File: 2025-01-January-GFG-POTD/January 14 - Container With Most Water/Solution.py
class Solution:
    def maxWater(self, arr):
        left = 0  # Left pointer
        right = len(arr) - 1  # Right pointer
        maxWater = 0  # To store the maximum water

        while left < right:
            # Calculate the water between the current pair of lines
            height = min(arr[left], arr[right])
            width = right - left
            currentWater = height * width

            # Update the maximum water if the current one is greater
            maxWater = max(maxWater, currentWater)

            # Move the pointer pointing to the shorter line inward
            if arr[left] < arr[right]:
                left += 1
            else:
                right -= 1

        return maxWater


if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        arr = list(map(int, input().strip().split()))
        obj = Solution()
        print(obj.maxWater(arr))
        print("~")
